scripts/knot_benchmark/src/benchmark/cases.py

```diff
@@ -0,0 +1,223 @@
+from __future__ import annotations
+
+import abc
+import enum
+import os
+import shutil
+from pathlib import Path
+import subprocess
+import sys
+import logging
+
+from benchmark import Command
+from benchmark.projects import Project
+
+
+class Benchmark(enum.Enum):
+    """Enumeration of the benchmarks to run."""
+
+    COLD = "cold"
+    """Cold check of an entire project without a cache present."""
+
+    WARM = "warm"
+    """Re-checking the entire project without any changes"."""
+
+    INCREMENTAL = "incremental"
+    """Incremental check between two commits."""
+
+
+def which_tool(name: str) -> Path:
+    tool = shutil.which(name)
+
+    assert (
+        tool is not None
+    ), f"Tool {name} not found. Run the script with `uv run <script>`."
+
+    return Path(tool)
+
+
+class Tool(abc.ABC):
+    def command(
+        self, benchmark: Benchmark, project: Project, venv: Venv
+    ) -> Command | None:
+        """Generate a command to benchmark a given tool."""
+        match benchmark:
+            case Benchmark.COLD:
+                return self.cold_command(project, venv)
+            case Benchmark.WARM:
+                return self.warm_command(project, venv)
+            case Benchmark.INCREMENTAL:
+                return self.incremental_command(project, venv)
+            case _:
+                raise ValueError(f"Invalid benchmark: {benchmark}")
+
+    @abc.abstractmethod
+    def cold_command(self, project: Project, venv: Venv) -> Command: ...
+
+    def warm_command(self, project: Project, venv: Venv) -> Command | None:
+        return None
+
+    def incremental_command(self, project: Project, venv: Venv) -> Command | None:
+        return None
+
+
+class Knot(Tool):
+    path: Path
+    name: str
+
+    def __init__(self, *, path: Path | None = None):
+        self.name = str(path) or "knot"
+        self.path = path or (
+            Path(__file__)
+            .joinpath(
+                "../../../../../",
+                "target",
+                "release",
+                "red_knot",
+            )
+            .resolve()
+        )
+
+        assert self.path.is_file(), f"Red Knot not found at '{self.path}'. Run `cargo build --release --bin red_knot`."
+
+    def cold_command(self, project: Project, venv: Venv) -> Command:
+        command = [str(self.path), "-v"]
+
+        assert len(project.include) < 2, "Knot doesn't support multiple source folders"
+
+        if directory := project.include[0]:
+            command.extend(["--current-directory", directory])
```

Comment:
`project.include` is typed as being `list[str]`, so `project.include[0]` will always be truthy unless it's the empty string `""`. Why are we specifically checking that `project.include[0]` isn't an empty string? Seems kinda unlikely to me that it would be

Reply:
Yeah, `IndexError` if it's an empty list and you try to index into it. And there's not `.get()` method for lists like there is in Rust, sadly

Reply:
I actually got an `IndexError` when I tried running the benchmark locally, so we do need to check the length of the list before indexing into it here:

```
~/dev/ruff (knot-mypy-pyright-benchmarks)⚡ [1] % cd scripts/knot_benchmark       
~/dev/ruff/scripts/knot_benchmark (knot-mypy-pyright-benchmarks)⚡ % uv run benchmark             
Using Python 3.12.4
Creating virtualenv at: .venv
   Built knot-benchmark @ file:///Users/alexw/dev/ruff/scripts/knot_benchmark
Installed 6 packages in 12ms
black (cold)
Benchmark 1: knot
  Time (mean ± σ):      71.3 ms ±   5.3 ms    [User: 58.6 ms, System: 9.1 ms]
  Range (min … max):    67.4 ms …  87.1 ms    41 runs
 
  Warning: Ignoring non-zero exit code.
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs.
 
Benchmark 2: Pyright
  Time (mean ± σ):      1.946 s ±  0.010 s    [User: 3.711 s, System: 0.131 s]
  Range (min … max):    1.930 s …  1.958 s    10 runs
 
  Warning: Ignoring non-zero exit code.
 
Benchmark 3: mypy
  Time (mean ± σ):      1.552 s ±  0.013 s    [User: 1.468 s, System: 0.081 s]
  Range (min … max):    1.539 s …  1.584 s    10 runs
 
  Warning: Ignoring non-zero exit code.
 
Summary
  knot ran
   21.78 ± 1.64 times faster than mypy
   27.30 ± 2.04 times faster than Pyright
black (warm)
Benchmark 1: mypy
  Time (mean ± σ):     314.5 ms ±   2.1 ms    [User: 271.6 ms, System: 40.1 ms]
  Range (min … max):   312.2 ms … 319.0 ms    10 runs
 
  Warning: Ignoring non-zero exit code.
 
Traceback (most recent call last):
  File "/Users/alexw/dev/ruff/scripts/knot_benchmark/.venv/bin/benchmark", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/alexw/dev/ruff/scripts/knot_benchmark/src/benchmark/run.py", line 136, in main
    if (command := suite.command(benchmark, project, venv))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alexw/dev/ruff/scripts/knot_benchmark/src/benchmark/cases.py", line 46, in command
    return self.cold_command(project, venv)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alexw/dev/ruff/scripts/knot_benchmark/src/benchmark/cases.py", line 88, in cold_command
    if directory := project.include[0]:
                    ~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```